### PR TITLE
✨ Reminder module

### PR DESCRIPTION
**Pull request by fatalerrorcoded**

### CODE REVIEW

File: bot/prisma/schema.prisma [hunk 1]
```diff
@@ -286,6 +289,20 @@ model Birthday {
   @@unique([userId, guildId])
 }
 
+// Reminders
+model Reminder {
+  id        String   @id @default(cuid())
+  createdAt DateTime @default(now())
+
+  userId     String
```
Comment (valivia) [inline bot/prisma/schema.prisma:297]: i'd group messageUrl, description and triggersAt together and user and userId as well
```prisma
  messageUrl String?
  description String?
  triggersAt  DateTime
  
  userId String
  user  User  @relation(fields: [userId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  ```

File: bot/src/modules/reminder/commands/reminder/list.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+export default SubCommand(
+
+  // Info
+  {
+    name: "list",
+    description: "List your reminders",
+
+    isGlobal: true,
+    arguments: [],
```
Comment (valivia) [inline bot/src/modules/reminder/commands/reminder/list.ts:15]: Makes more sense to remove the unused `arguments` property

File: bot/src/modules/reminder/commands/reminder/list.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    const embed = embedTemplate()
+      .setTitle("Reminder List");
+
+    const reminders = await db.reminder.findMany({ where: { userId: msg.user.id } });
+    if (reminders.length === 0) {
+      embed.setDescription("You have no reminders");
+      return { embeds: [embed] };
+    }
+
+    reminders.sort((a, b) => a.triggersAt.getTime() - b.triggersAt.getTime());
```
Comment (valivia) [inline bot/src/modules/reminder/commands/reminder/list.ts:34]: I believe prisma can do the sorting as well?

File: bot/src/modules/reminder/commands/reminder/remove.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    const reminder = await db.reminder.findFirst({
+      where: { userId: msg.user.id },
+      orderBy: { triggersAt: "asc" },
+      skip: position - 1,
+    });
+
+    if (!reminder)
+      return { embeds: [ failEmbedTemplate("Reminder does not exist") ] };
+
+    deleteReminder(reminder.id);
```
Comment (valivia) [inline bot/src/modules/reminder/commands/reminder/remove.ts:49]: async?

File: bot/src/modules/reminder/lib/createReminder.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  const index = localState.reminders.findIndex((reminderRef) => reminderRef.triggersAt.getTime() > triggersAt.getTime());
+  if (index === -1) {
+    // if index is -1, that means that all existing reminders occur before this reminder, thus this reminder needs to be last
+    localState.reminders.push(toInsert);
+  } else {
+    // there is a reminder that occurs after this reminder, thus we insert this reminder before that reminder
+    localState.reminders.splice(index, 0, toInsert);
+  }
+
+  return entry.id;
```
Comment (valivia) [inline bot/src/modules/reminder/lib/createReminder.ts:28]: why only return the id?

File: bot/src/modules/reminder/lib/loadRemindersFromDb.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { state } from "@app";
+import { ReminderRef } from "../structs/reminderRef";
+
+const db = state.db;
+
+export async function loadRemindersFromDb(): Promise<ReminderRef[]> {
+  const reminders = await db.reminder.findMany({ select: { id: true, triggersAt: true } });
+  reminders.sort((a, b) => a.triggersAt.getTime() - b.triggersAt.getTime());
```
Comment (valivia) [inline bot/src/modules/reminder/lib/loadRemindersFromDb.ts:8]: prisma sort?

File: bot/src/modules/reminder/cron/trigger.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        continue;
+      }
+
+      const embed = embedTemplate()
+        .setTitle("Reminder")
+        .setDescription(`Here is your reminder in ${reminder.messageUrl} from <t:${Math.round(reminder.createdAt.getTime() / 1000)}>`);
+
+      if (reminder.description)
+        embed.addFields({ name: "Description", value: reminder.description });
+
```
Comment (valivia) [inline bot/src/modules/reminder/cron/trigger.ts:43]: I'd probably add a debug log when it succeeds

File: bot/src/modules/reminder/commands/reminder/set.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+    const timeoutFormatted = msToString(timeoutMs);
+    const embed = embedTemplate()
+      .setTitle("Reminder set")
+      .setDescription(`You will be reminded in ${timeoutFormatted} on the <t:${Math.round(timeoutDate.getTime() / 1000)}>`)
+      .setFooter({ text: "You will be reminded via DM, make sure you can receive DMs from the bot!" });
+
+    return {
+      embeds: [embed],
+      callback: async (interaction) => {
```
Comment (valivia) [inline bot/src/modules/reminder/commands/reminder/set.ts:65]: I'd probably move this into its own function below the main command code

**Review by valivia — CHANGES_REQUESTED** [review]

File: bot/src/modules/reminder/lib/createReminder.ts (hunk 5 above)
Comment (fatalerrorcoded) [reply bot/src/modules/reminder/lib/createReminder.ts:28]: what return type should i use then if i am to return a prisma type :trollface:

**Review by valivia — APPROVED** [review]

**Pull request merged**